File: claude_exp/xv3/router/upstream.py
import socket
import threading
import time
from typing import Optional, Tuple

from shared.framing import read_message, write_message

UpstreamConn = Tuple[socket.socket, tuple, threading.Lock]


class UpstreamServer:
    """Listen on cfg.port. Created once outside the session loop (survives reconnects)."""

    def __init__(self, cfg):
        self.cfg = cfg
        self._sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self._sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self._sock.bind((cfg.host, cfg.port))
        self._sock.listen(5)
        self._sock.settimeout(1.0)

    def accept(self, stop_event) -> Optional[UpstreamConn]:
        while not stop_event.is_set():
            try:
                conn, addr = self._sock.accept()
            except socket.timeout:
                continue
            except OSError:
                return None
            return conn, addr, threading.Lock()
        return None

    def close(self):
        self._sock.close()


class UpstreamClient:
    """Connect out to cfg.host:cfg.port, retrying every cfg.retry_seconds."""

    def __init__(self, cfg):
        self.cfg = cfg

    def connect(self, stop_event) -> Optional[UpstreamConn]:
        # Polls stop_event.is_set() every second rather than calling stop_event.wait()
        # directly, so callers can pass any duck-typed "is_set()"-only event (e.g. a
        # combinator of multiple events) in place of a real threading.Event.
        while not stop_event.is_set():
            try:
                sock = socket.create_connection((self.cfg.host, self.cfg.port), timeout=5)
                return sock, (self.cfg.host, self.cfg.port), threading.Lock()
            except OSError:
                waited = 0.0
                while waited < self.cfg.retry_seconds and not stop_event.is_set():
                    time.sleep(min(1.0, self.cfg.retry_seconds - waited))
                    waited += 1.0
        return None


def read_upstream(conn, cfg) -> bytes:
    return read_message(conn, cfg.framing.to_dict())


def write_upstream(conn, data: bytes, cfg) -> None:
    write_message(conn, data, cfg.framing.to_dict())

File: claude_exp/routers/router_java/src/main/java/com/xv6/router/Upstream.java
package com.xv6.router;

import com.xv6.shared.Framing;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.SocketTimeoutException;
import java.util.concurrent.locks.ReentrantLock;
import java.util.function.BooleanSupplier;

/**
 * Port of router_py's router/upstream.py. {@code shouldStop} plays the role of Python's combined
 * stop_event/reconnect_event ({@code _OrEvent}) - callers pass e.g. {@code orEvent::isSet}.
 * Waits below poll every 200ms rather than blocking on a single condition variable (unlike
 * Python's threading.Event.wait, which wakes immediately when set); this trades a little latency
 * for a much simpler implementation, acceptable given this project's multi-second retry/reestablish
 * intervals.
 */
public final class Upstream {

    private Upstream() {
    }

    public static byte[] readUpstream(Socket conn, UpstreamConfig cfg) throws IOException {
        return Framing.readMessage(conn, cfg.framing());
    }

    public static void writeUpstream(Socket conn, byte[] data, UpstreamConfig cfg) throws IOException {
        Framing.writeMessage(conn, data, cfg.framing());
    }

    /** Listens on cfg.port. Created once outside the session loop (survives reconnects). */
    public static final class UpstreamServer {
        private final ServerSocket serverSocket;

        public UpstreamServer(UpstreamConfig cfg) throws IOException {
            serverSocket = new ServerSocket();
            serverSocket.setReuseAddress(true);
            serverSocket.bind(new InetSocketAddress(cfg.port()));
            serverSocket.setSoTimeout(1000);
        }

        /** Loops with a 1-second accept timeout; returns null on stop or a hard socket error. */
        public UpstreamConn accept(BooleanSupplier shouldStop) {
            while (!shouldStop.getAsBoolean()) {
                try {
                    Socket sock = serverSocket.accept();
                    return new UpstreamConn(sock, sock.getRemoteSocketAddress(), new ReentrantLock());
                } catch (SocketTimeoutException timeout) {
                    // loop, re-check shouldStop
                } catch (IOException e) {
                    return null;
                }
            }
            return null;
        }

        public void close() {
            try {
                serverSocket.close();
            } catch (IOException ignored) {
            }
        }
    }

    /** Connects out to cfg.host:cfg.port, retrying every cfg.retrySeconds(). */
    public static final class UpstreamClient {
        private final UpstreamConfig cfg;

        public UpstreamClient(UpstreamConfig cfg) {
            this.cfg = cfg;
        }

        public UpstreamConn connect(BooleanSupplier shouldStop) {
            while (!shouldStop.getAsBoolean()) {
                try {
                    Socket sock = new Socket();
                    sock.connect(new InetSocketAddress(cfg.host(), cfg.port()), 5000);
                    return new UpstreamConn(sock, sock.getRemoteSocketAddress(), new ReentrantLock());
                } catch (IOException e) {
                    if (!interruptibleWait(cfg.retrySeconds() * 1000L, shouldStop)) {
                        return null;
                    }
                }
            }
            return null;
        }

        private boolean interruptibleWait(long totalMillis, BooleanSupplier shouldStop) {
            long elapsed = 0;
            while (elapsed < totalMillis) {
                if (shouldStop.getAsBoolean()) {
                    return false;
                }
                try {
                    Thread.sleep(Math.min(200, totalMillis - elapsed));
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    return false;
                }
                elapsed += 200;
            }
            return true;
        }
    }
}
